code/modules/mob/living/ventcrawling.dm

```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
+			forceMove(A.loc)
+			REMOVE_TRAIT(src, TRAIT_MOVE_VENTCRAWLING, VENTCRAWLING_TRAIT)
+			update_pipe_vision()
+		//Entrance here
+		else
+			var/datum/pipeline/vent_parent = A.parents[1]
+			if(vent_parent && (vent_parent.members.len || vent_parent.other_atmosmch))
+				visible_message("<span class='notice'>[src] begins climbing into the ventilation system...</span>" ,"<span class='notice'>You begin climbing into the ventilation system...</span>")
+				if(!do_after(src, 25, target = A))
+					return
+				if(!client)
 					return
+				visible_message("<span class='notice'>[src] scrambles into the ventilation ducts!</span>","<span class='notice'>You climb into the ventilation ducts.</span>")
+				forceMove(A)
+				ADD_TRAIT(src, TRAIT_MOVE_VENTCRAWLING, VENTCRAWLING_TRAIT)
+				update_pipe_vision()
+			else
+				to_chat(src, "<span class='warning'>This ventilation duct is not connected to anything!</span>")
 
-			visible_message("<span class='notice'>[src] scrambles into the ventilation ducts!</span>","<span class='notice'>You climb into the ventilation ducts.</span>")
-			forceMove(vent_found)
-	else
-		to_chat(src, "<span class='warning'>This ventilation duct is not connected to anything!</span>")
 
 /mob/living/simple_animal/slime/handle_ventcrawl(atom/A)
 	if(buckled)
 		to_chat(src, "<i>I can't vent crawl while feeding...</i>")
 		return
 	return ..()
 
-
-/mob/living/proc/add_ventcrawl(obj/machinery/atmospherics/starting_machine)
-	if(!istype(starting_machine) || !starting_machine.can_see_pipes())
-		return
-	var/list/totalMembers = list()
-
-	for(var/datum/pipeline/P in starting_machine.returnPipenets())
-		totalMembers += P.members
-		totalMembers += P.other_atmosmch
-
-	if(!totalMembers.len)
-		return
-
-	if(client)
-		for(var/X in totalMembers)
-			var/obj/machinery/atmospherics/A = X //all elements in totalMembers are necessarily of this type.
-			if(in_view_range(client.mob, A))
-				if(!A.pipe_vision_img)
-					A.pipe_vision_img = image(A, A.loc, layer = ABOVE_HUD_LAYER, dir = A.dir)
-					A.pipe_vision_img.plane = ABOVE_HUD_PLANE
-				client.images += A.pipe_vision_img
-				pipes_shown += A.pipe_vision_img
-	ADD_TRAIT(src, TRAIT_MOVE_VENTCRAWLING, VENTCRAWLING_TRAIT)
-
-/mob/living/proc/remove_ventcrawl()
-	if(client)
-		for(var/image/current_image in pipes_shown)
-			client.images -= current_image
-	pipes_shown.len = 0
-	REMOVE_TRAIT(src, TRAIT_MOVE_VENTCRAWLING, VENTCRAWLING_TRAIT)
-
-
-
-
-//OOP
-/atom/proc/update_pipe_vision(atom/new_loc = null)
-	return
-
-/mob/living/update_pipe_vision(atom/new_loc = null)
-	. = loc
-	if(new_loc)
-		. = new_loc
-	remove_ventcrawl()
-	add_ventcrawl(.)
+/mob/living/proc/update_pipe_vision()
+	if(HAS_TRAIT(src, TRAIT_MOVE_VENTCRAWLING) && istype(src.loc, /obj/machinery/atmospherics) && movement_type & VENTCRAWLING)
+		var/list/totalMembers = list()
+		var/obj/machinery/atmospherics/current_location = src.loc
+		for(var/datum/pipeline/P in current_location.returnPipenets())
```

Comment:
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names